fix: race condition causing exec mode to hang with --image flag

jet-ds opened the pull request:
## What

Fixes a race condition and dropped events in `codex exec` when using the `--image` flag that causes the process to hang indefinitely.

## Why

**Race Condition**: When images are submitted via `codex exec --json --image <path>`, two threads compete for the same event stream:
1. Background thread (line 298): `conversation.next_event()` → forwards to channel `rx`
2. Main thread (line 330): `conversation.next_event()` → waits for `TaskComplete`

Both consume from the same `async_channel::Receiver` (cloned via `conversation.clone()`). When the background thread wins the race (which happens deterministically), the main thread hangs forever.

**Dropped Events**: Additionally, even after fixing the race by reading from `rx`, events were not being processed during the wait. This silently dropped the assistant's response to image uploads.

**Impact**: Makes `codex exec --image` completely unusable. Affects any programmatic integration using images (CI/CD, API wrappers, etc.).

## How

The fix has two parts:
1. Changed line 330 to read from `rx.recv()` instead of `conversation.next_event()` (fixes the race)
2. Added `event_processor.process_event(event)` inside the loop (fixes dropped events)

This ensures events are both consumed from the correct channel and properly processed.

## Credit

Race condition diagnosed by @AdsQnn in #5773.
Dropped events issue caught by @etraut-openai during review.

## Testing

- ✅ `cargo fmt --all -- --config imports_granularity=Item` - passed
- ✅ `cargo clippy --tests` - passed (no warnings)
- ✅ `cargo test --package codex-exec` - all tests passed
- ✅ Manual testing: `codex exec --json --image <path>` now completes successfully

Tested on macOS ARM (aarch64-apple-darwin).

Fixes #5773

jet-ds committed:
Fix race condition in exec mode image handling

When images are submitted in exec mode, two threads compete for events
from the same conversation stream: the background event-forwarding thread
and the main thread waiting for TaskComplete. This causes the main thread
to hang indefinitely when the background thread wins the race.

Fix by reading from the rx channel instead of calling conversation.next_event()
directly, ensuring all events flow through the background thread's channel.

Credit to @AdsQnn for diagnosing this issue in #5773.

Signed-off-by: Jet Sanchez <jetsanchezzz@gmail.com>

etraut-openai committed:
Merge branch 'main' into fix/exec-image-race-condition

AdsQnn commented:
The Rust fix resolves a race condition, but it introduces a new problem.

Problem with my solution:

 ❌ Incorrect fix:
```
while let Some(event) = rx.recv().await {
    if event.id == initial_images_event_id
        && matches!(event.msg, EventMsg::TaskComplete(..)) {
        break;
    }
    // ❌ The event disappears! It never reaches the event_processor
}
```

The main loop will never see these events:
```
while let Some(event) = rx.recv().await {
    event_processor.process_event(event);  // Won’t receive events from the loop above
}
```

Effect: All events between sending the images and TaskComplete (e.g. assistant responses about the images)
are lost — they never reach the event_processor.

Correct solution:

✅ Read from rx AND process every event
```
while let Some(event) = rx.recv().await {
    // Check if this is the completion event for images
    let is_images_complete = event.id == initial_images_event_id
        && matches!(event.msg, EventMsg::TaskComplete(..));

    // ALWAYS process the event (so nothing gets lost)
    event_processor.process_event(event);

    // If this was the TaskComplete for images, stop waiting
    if is_images_complete {
        break;
    }
}
```

Key difference: Every event MUST go through the event_processor,
even while waiting for a specific event — otherwise, user-facing output is lost.

Good point — it might help with something.
However, I’m only checking this occasionally, so it needs further analysis.
I’m not able to fully verify it right now :))
Maybe the author will consider committing it :)

jet-ds committed:
Fix race condition and dropped events in exec mode image handling

When images are submitted in exec mode, two threads compete for events
from the same conversation stream: the background event-forwarding thread
and the main thread waiting for TaskComplete. This causes the main thread
to hang indefinitely when the background thread wins the race.

Additionally, events consumed during the wait were not being processed,
silently dropping the assistant's response to image uploads.

Fix by reading from the rx channel (avoiding the race) and processing
each event before checking if it is the TaskComplete signal.

Credit to @AdsQnn for diagnosing the race condition in #5773.
Thanks to @etraut-openai for catching the dropped events issue.

Signed-off-by: Jet Sanchez <jetsanchezzz@gmail.com>

jet-ds committed:
Merge remote-tracking branch 'upstream/main' into fix/exec-image-race-condition

jet-ds committed:
Merge remote-tracking branch 'upstream/main' into fix/exec-image-race-condition

etraut-openai committed:
Merge branch 'main' into fix/exec-image-race-condition
